logs/gateway-ingress/src/publish.rs

```diff
@@ -0,0 +1,852 @@
+        event_tx: mpsc::Sender<Event>,
+    ) {
+        raw_event_stream
+            .for_each(|event| async {
+                if let Err(send_err) = event_tx.try_send(event) {
+                    slog::warn!(
+                        self.logger,
+                        "sending raw gateway event into shared channel failed; dropping";
+                        "error" => ?send_err
+                    );
+                }
+            })
+            .await;
+    }
+
+    /// Asynchronously watches the queue length to ensure
+    /// that it doesn't exceed a warning threshold.
+    /// This is useful for reporting before the queue starts dropping events.
+    async fn watch_queue_length(&self, check_length: impl Fn() -> usize) {
+        let mut interval = tokio::time::interval(self.config.raw_events.watch_period);
+        loop {
+            interval.tick().await;
+            let current_length = check_length();
+            if current_length > self.config.raw_events.warn_threshold {
+                slog::warn!(
+                    self.logger,
+                    "current queue length exceeds warning threshold";
+                    "current_queue_length" => current_length,
+                    "warn_threshold" => self.config.raw_events.warn_threshold
+                );
+            }
+        }
+    }
+
+    /// Publishes a single event to the RabbitMQ message queue.
+    /// Blocks until the event can be successfully published,
+    /// potentially indefinitely.
+    /// This can cause events to start being dropped on the bounded queue.
+    async fn publish_event(&self, event: Event) {
+        // Provision an ID and note the timestamp immediately
+        let timestamp = time::millisecond_ts();
+        let id = self.id_provisioner.with_ts(timestamp);
+        let logger = self.logger.new(slog::o!(
+            "event_timestamp" => timestamp,
+            "event_id" => id,
+        ));
+
+        // Create the `GatewayEvent` from the raw event
+        let gateway_event =
+            if let Some(gateway_event) = convert_raw_event(event, timestamp, id, &logger) {
+                gateway_event
+            } else {
+                return;
+            };
+
+        let logger = self.logger.new(slog::o!(
+            "guild_id" => gateway_event.guild_id,
+        ));
+
+        // Serialize the event to raw bytes
+        let mut raw_bytes = Vec::with_capacity(gateway_event.encoded_len());
+        if let Err(encode_err) = gateway_event.encode(&mut raw_bytes) {
+            slog::warn!(
+                logger,
+                "could not encode gateway event to bytes";
+                "error" => ?encode_err,
+                "event" => ?gateway_event
+            );
+            return;
+        }
+
+        // Print out the raw bytes for debugging
+        slog::debug!(
+            logger,
+            "raw event bytes";
+            "event_bytes" => base64::encode(&raw_bytes),
+        );
```

Comment:
It's only used when the log level is debug, so it doesn't do anything in production. Because when developing locally the bot won't be in many guilds, this shouldn't be very noisy (and might be helpful in debugging events being dropped due to buggy code flagging them as unnecessary).